presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EnforceRequiredJoinMethods.java

```diff
@@ -0,0 +1,143 @@
+import static com.facebook.presto.sql.planner.optimizations.ScalarQueryUtil.isScalar;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static java.util.Objects.requireNonNull;
+
+public class EnforceRequiredJoinMethods
+        implements PlanOptimizer
+{
+    private boolean isDeleteQuery;
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(session), plan);
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+
+        public Rewriter(Session session)
+        {
+            this.session = session;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode leftRewritten = context.rewrite(node.getLeft(), context.get());
+            PlanNode rightRewritten = context.rewrite(node.getRight(), context.get());
+            JoinNode.Method targetJoinMethod = getTargetJoinMethod(node);
+            return new JoinNode(
+                    node.getId(),
+                    node.getType(),
+                    leftRewritten,
+                    rightRewritten,
+                    node.getCriteria(),
+                    node.getFilter(),
+                    node.getLeftHashSymbol(),
+                    node.getRightHashSymbol(),
+                    Optional.of(targetJoinMethod));
+        }
+
+        @Override
+        public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Void> context)
+        {
+            PlanNode sourceRewritten = context.rewrite(node.getSource(), context.get());
+            PlanNode filteringSourceRewritten = context.rewrite(node.getFilteringSource(), context.get());
+            SemiJoinNode.Method targetJoinMethod = getTargetSemiJoinMethod(isDeleteQuery);
+            return new SemiJoinNode(
+                    node.getId(),
+                    sourceRewritten,
+                    filteringSourceRewritten,
+                    node.getSourceJoinSymbol(),
+                    node.getFilteringSourceJoinSymbol(),
+                    node.getSemiJoinOutput(),
+                    node.getSourceHashSymbol(),
+                    node.getFilteringSourceHashSymbol(),
+                    Optional.of(targetJoinMethod));
+        }
+
+        @Override
+        public PlanNode visitDelete(DeleteNode node, RewriteContext<Void> context)
+        {
+            // For delete queries, the TableScan node that corresponds to the table being deleted must be collocated with the Delete node,
```

Comment:
There's a test in AbstractTestDistributedQueries in the delete using a subquery section.